design/extension.go

```diff
@@ -0,0 +1,963 @@
+package design
+
+import (
+	"time"
+
+	. "goa.design/goa/v3/dsl"
+)
+
+var _ = Service("extension", func() {
+	Description("The extension registry service interfaces with the FuseML Extension Registry.")
+
+	Method("registerExtension", func() {
+		Description("Register an external utility as a FuseML extension with the FuseML extension registry.")
+
+		Payload(Extension, "Extension registration request")
+
+		Error("BadRequest", func() {
+			Description("If the extension does not have the required fields, should return 400 Bad Request.")
+		})
+		Error("Conflict", func() {
+			Description("If an extension with the same name already exists, should return 409 Conflict.")
+		})
+
+		Result(Extension)
+
+		HTTP(func() {
+			POST("/extensions")
+			Response(StatusCreated)
+			Response("BadRequest", StatusBadRequest)
+			Response("Conflict", StatusConflict)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("Conflict", CodeAlreadyExists)
+		})
+	})
+
+	Method("getExtension", func() {
+		Description("Retrieve information about an extension.")
+
+		Payload(func() {
+			Field(1, "id", String, "Extension identifier", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("kfserving-001")
+			})
+			Required("id")
+		})
+
+		Error("NotFound", func() {
+			Description("If there is no extension with the given ID, should return 404 Not Found.")
+		})
+
+		Result(Extension)
+
+		HTTP(func() {
+			GET("/extensions/{id}")
+			Response(StatusOK)
+			Response("NotFound", StatusNotFound)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+		})
+	})
+
+	Method("listExtensions", func() {
+		Description("List extensions registered in FuseML")
+
+		Result(ArrayOf(Extension), "Return all registered extensions.")
+
+		HTTP(func() {
+			GET("/extensions")
+			Response(StatusOK)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+		})
+	})
+
+	Method("updateExtension", func() {
+		Description("Update an extension registered in FuseML")
+
+		Payload(Extension, "Extension update request")
```

Comment:
I understand that you get the extension ID from the extension struct passed as parameter, but I am still not sure about this.

Isn't it common to pass the ID explicitly during the update operations, so that the object to update is clearly identified and only the _other_ parameter holds the data which should be used for the actual update?

What if, for example, you want to change the ID itself? Your methods will not find the original objects...

Reply:
I've been battling with this same concept for a couple of days now. Yes, you're right, they should be different things: the ID should be a URL parameter or part of the URL path, the rest of the attributes should be in the request body.

Given that the extension registry implementation doesn't allow changing the ID attribute during an update (and I don't intend to add support for that because it complicates things unnecessarily), I thought reusing the body attributes would provide for a cleaner implementation, which it does.

I can try switching to use parameters _in addition_ to the ID attributes in the body. The one thing I can't do is to remove the ID attributes from the body, because that would mean having to create another set of objects like Extension and ExtensionService that don't include those IDs, which would mean having to duplicate all the boilerplate code that is required to convert them to/from domain.

Reply:
Ok, it looks like I need the IDs to be in the payload too, even if they're modeled as parameters. Besides, I don't need to model them as parameters because they are part of the URL path anyway.

This only leaves me with the idea of having to separate the IDs from the regular payload, which means the request body will have to look like this:

```
# curl -X PUT --data-binary @update-credentials001.yaml -H "Content-Type: application/x-yaml" http://localhost:8000/extensions/mlflow-0001/services/mlflow-store/credentials/default-s3-account
credentials:
  scope: project
  projects:
    - project-001
  configuration:
    AWS_ACCESS_KEY_ID: oT0SfbJPEu6kUbUKsH24
    AWS_SECRET_ACCESS_KEY: GiZff8KqS5xWQ4eagRujh1tDcbQyRP0bEJSBOfs5
```

I really don't see that this will make a difference. The user could still include the IDs in the update request, e.g.:

```
id: default-s3-account
extension_id: mlflow-0001
service_id: mlflow-store
credentials:
  scope: project
  projects:
    - project-001
  configuration:
    AWS_ACCESS_KEY_ID: oT0SfbJPEu6kUbUKsH24
    AWS_SECRET_ACCESS_KEY: GiZff8KqS5xWQ4eagRujh1tDcbQyRP0bEJSBOfs5
```

At least with the current approach, the IDs that are included in the request are automatically overwritten by the REST service, so I don't need to worry about complications.

I would like to stick with the current implementation, if that's the same to you.

Reply:
Hm, I think the approach with identifiers being separate (not necessary extracted from the payload) is nicer, but I guess I do not really mind.

Anyway, please take a look at the other Update related issue which I mention here https://github.com/fuseml/fuseml-core/pull/93#discussion_r691183901